README.md

```diff
@@ -1,10 +1,260 @@
+Adminrouter repository includes a test harness that is meant to make
+testing easier and in some cases - possible. It's written in Python and
+uses pytest fixtures and custom modules to mock out all relevant DC/OS
+features and control Nginx startup and termination.
+
+All the tests are executed in a Docker container which is controlled by the
+Makefile. Inside the container pytest command is started which in turn pulls
+in all the relevant fixtures in, such as syslog mock, mocker (DC/OS endpoints
+mock), dns mock, etc... Finally, an Nginx is spawned using the configuration
+bind-mounted from the developer's repository. Tests may launch Nginx multiple
+times, in different configuration, depending on what is needed. After the
+tests runner finishes, all the processes and the environment is cleaned up
+by pytest.
+
+Below, there is a general overview of all the components of the test harness.
+More detailed documentation can be found in docstrings and comments in the code
+itself. If in doubt, it also may be very helpful to look for examples among
+existing tests.
+
+### Quickstart
+In order to execute all the test, just issue:
+
+	make test
+
+In order have fine grained control over pytest command, execute:
+
+	make shell
+
+This will launch an interactive environment inside the container that has all
+the dependencies installed. The developer may now launch the `pytest` command
+the way it's necessary, debug the environment and temporarily add/change
+the dependencies.
+
+### Makefile
+Makefile provides an easy way to start the testing environment without the need
+to worry about the correct docker commands. It's core concept is `adminrouter-devkit`
+container which is contains all the dependencies that are needed to run
+Adminrouter, and inside which all tests-related commands are run.
+
+It exposes a couple of targets:
+* `make clean` - remove all containers created by the test harness. It does not
+  remove the images themselves though as the layer cache may be useful later
+  on and the user may remove them themselves.
+* `make devkit` - creates the `adminrouter-devkit` container. By default other
+   targets execute this step automatically if devkit container image does not
+   exist yet.
+* `make update-devkit` - updates `adminrouter-devkit`. Should be run every time
+   the Dockerfile or its dependencies change.
+* `make tests` - launch all the tests. Worth noting is the fact that mccabe
+  complexity of the code is also verified and an error is raised if it's equal
+  or above 10.
+* `make shell` - launch an interactive shell within the devkit container. Should
+  be used in case when fine grained control of the tests is necessary or
+  during debugging.
+* `make flake8` - launch flake8 which will check all the tests and test-harness
+  files by default.
+
+### Docker container
+As mentioned earlier, all the commands are executed inside the `adminrouter-devkit`
+container`. It follows the same build process for Nginx that happens during
```

Comment:
I didn't mean to offend -- this additional tick changed syntax highlighting and rendering of a _huge_ chunk of the entire README and made review actually pretty hard.